streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -659,13 +665,12 @@ void runOnce() {
             }
         }
 
-        // we can always let changelog reader try restoring in order to initialize the changelogs;
-        // if there's no active restoring or standby updating it would not try to fetch any data
-        changelogReader.restore();
-
-        // TODO: we should record the restore latency and its relative time spent ratio after
-        //       we figure out how to move this method out of the stream thread
-        advanceNowAndComputeLatency();
+        // check if restore thread has encountered TaskCorrupted exception; if yes
+        // rethrow it to trigger the handling logic
+        final TaskCorruptedException e = restoreThread.nextCorruptedException();
```

Comment:
Might be nice to gather and handle all TaskCorruptedExceptions at once rather than one per loop like this, especially since each one likely involves committing all tasks (and I would imagine that with EOS, when we get one TaskCorrupted we are likely to  also have more). That can definitely be follow-on work, just putting it out there

Reply:
Make sense, will try to address this in this PR.

Reply:
@ableegoldman I've not addressed this comment but it's on my next PR to do it, this one has become too large and hence I'm only keeping necessary changes to get correctness.